{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "13f15429-2411-4358-a378-b2e14e87bb30",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import SVG"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4faff53-f01c-4a0d-8c26-f15b83815d82",
   "metadata": {},
   "source": [
    "# Metagenomics\n",
    "\n",
    "## Steps of the analysis\n",
    "\n",
    "* QC of raw reads with [fastqc](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)\n",
    "* Read trimming with [fastp](https://github.com/OpenGene/fastp)\n",
    "    * Removes adapters, drops reads with average PHRED quality  below 20\n",
    "    * Does QC of trimmed reads\n",
    "* Mapping reads to reference with [minimap2](https://github.com/lh3/minimap2)\n",
    "* Filter alignment files with samtools view\n",
    "    * Removes reads that are not mapped in propper paris\n",
    "    * Removes reads with a mapping PHRED score below 60 (in minimap2 this is equilivant to 30)\n",
    "* Call variants with [freebayes](https://github.com/freebayes/freebayes)\n",
    "    * Allele needs to be present 3 times\n",
    "    * Minimum allele fraction is 0.05\n",
    "    * Minimum 30 PHRED mapping quality (redundant because of filtering)\n",
    "    * Minimum 20 PHRED base quality\n",
    "* Annotate alleles with [SnpEFF](https://pcingola.github.io/SnpEff/snpeff/build_db/#step-2-option-2-building-a-database-from-genbank-files)\n",
    "* Filter alleles with custom python script [vcf_exporter.py](https://github.com/sulheim/leakage/blob/eric/code/meta_sequencing/results/vcf_exporter.py)\n",
    "    * All frequencies\n",
    "        * Minimum allele frequency: 0.05\n",
    "        * Minimum PHRED qual calculated by freebayes 30\n",
    "        * Minimum allele count: 5\n",
    "        * Minimum depth at allele site: 30\n",
    "    * Fixed alleles:\n",
    "        * Minimum allele frequency: 0.9\n",
    "        * Minimum PHRED qual calculated by freebayes 30\n",
    "        * Minimum depth at allele site: 30\n",
    "\n",
    "## File- and directory strucutre\n",
    "\n",
    "There was a disk quota issue so the processing was done on `/work/FAC/FBM/DMF/smitri/evomicrocomm/seq_snorre/data/meta_sequencing`.\n",
    "`references_sequencing/` contains FASTA and annotation files, `meta_sequencing/` contains all files from the samples.  \n",
    "The code is available in this repo under `../../code/meta_sequencing/` where mainly the Snakefile is important.  \n",
    "`vcf_exporter.py` grabs the files from freebayes and filters the alleles. The output is a csv that I dump under the `../code/meta_sequencing/results` in different folders:\n",
    "* `all_variants`:\n",
    "    * All variants outputted by freebayes (note there is already some default filtering)\n",
    "* `all_filtered_variants`:\n",
    "    * All filtered variants according to `vcf_exporter.py` filter described above\n",
    "* `fixed_filtered_variants`:\n",
    "    * Fixed alleles filtered according to `vcf_exporter.py` described above\n",
    "      \n",
    "The `csv` contains following columns:  \n",
    "`chrom,pos,qual,depth,freq,alt,alt_count,ref,type,len,eff,gene,product,linegroup,sample`\n",
    "* `freq`: Allele frequency\n",
    "* `alt`: Nucleotite,deletions,or insertion in allele\n",
    "* `alt_count`: Number of reads with the observed allele\n",
    "* `ref`: Nucleotide or sequence in reference\n",
    "* `type`: Type of mutation\n",
    "* `eff`: Effect if CDS on translated feature\n",
    "* `linegroup`: Unique identifiers for each allele (also unique across samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08a7f213-9252-44a6-a561-95c0c15171b9",
   "metadata": {},
   "source": [
    "## Some quality checks\n",
    "\n",
    "We can look at the area of the knockout genes AceE and SucB to check if no strains were confused during the processing. Everything looks good!\n",
    "\n",
    "### AceE (Position 119504-122167)\n",
    "\n",
    "![meta_acee](../screenshots/meta_alignments_AceE.png)\n",
    "\n",
    "### SucB (Position 756978-758195)\n",
    "\n",
    "![meta_sucb](../screenshots/meta_alignments_SucB.png)\n",
    "\n",
    "## Coverage\n",
    "![coverage](../screenshots/fig1_coverage.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0784b1cd-3de1-4e89-81ed-f7bbc0d7e2ba",
   "metadata": {},
   "source": [
    "## Frequencies of alleles across samples (each dot is one allele)\n",
    "\n",
    "![allel_freq](../screenshots/fig2_frequencies.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9c9f22b-694c-4a72-80d7-762b24e34f0c",
   "metadata": {},
   "source": [
    "## Fixed alleles (frequency >= 0.9) across samples\n",
    "\n",
    "***ynfM* Putative arabinose efflux transporter fixed in all AceE samples**\n",
    "\n",
    "Two not annotated SNPs across all samples for SucB\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "ba2d7c35-7e43-4149-b7ad-0696ce4486bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>gene</th>\n",
       "      <th>product</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>1664690</td>\n",
       "      <td>ynfM</td>\n",
       "      <td>putative arabinose efflux transporter</td>\n",
       "      <td>AceE_M4_D44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>2719426</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AceE_M4_D44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         pos  gene                                product       sample\n",
       "141  1664690  ynfM  putative arabinose efflux transporter  AceE_M4_D44\n",
       "142  2719426   NaN                                    NaN  AceE_M4_D44"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../../code/meta_sequencing/results/fixed_filtered_variants/all_samples.filtered.csv')[['pos','gene','product','sample']]\n",
    "aceE = df[df['sample'] == 'AceE_M4_D44']\n",
    "aceE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "337ec64e-7742-4c0d-94b2-1e70fd083c7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>gene</th>\n",
       "      <th>product</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>447145</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SucB_M5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>2719426</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SucB_M5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         pos gene product   sample\n",
       "143   447145  NaN     NaN  SucB_M5\n",
       "144  2719426  NaN     NaN  SucB_M5"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../../code/meta_sequencing/results/fixed_filtered_variants/all_samples.filtered.csv')[['pos','gene','product','sample']]\n",
    "SucB = df[df['sample'] == 'SucB_M5']\n",
    "SucB"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9eb3006e-2724-40eb-95c4-9293516a8e5c",
   "metadata": {},
   "source": [
    "## Alleles present across samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "dd484115-5fb5-46c3-9b2e-ad76d09ada7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nan\n",
      "isocitrate dehydrogenase kinase/phosphatase\n",
      "LPS(HepIII)-glucuronic acid glycosyltransferase\n"
     ]
    }
   ],
   "source": [
    "aceE_samples = ['AceE_M2_D44','AceE_M3_D44','AceE_M4_D44']\n",
    "sucB_samples = ['SucB_M5','SucB_M6','SucB_M7']\n",
    "all_samples = aceE_samples + sucB_samples\n",
    "\n",
    "def parallel_products(samples,min_count):\n",
    "    dfs = [pd.read_csv('../../code/meta_sequencing/results/all_filtered_variants/' + sample + '.filtered.csv') for sample in samples]\n",
    "    products = []\n",
    "    for df in dfs:\n",
    "        products += df['product'].to_list()\n",
    "    product_dict = {key:0 for key in set(products)}\n",
    "    \n",
    "    for df in dfs:\n",
    "        for product in product_dict.keys():\n",
    "            if product in df['product'].to_list():\n",
    "                product_dict[product] += 1\n",
    "    \n",
    "    for product,count in product_dict.items():\n",
    "        if count >= min_count:\n",
    "            print(product)\n",
    "\n",
    "parallel_products(sucB_samples,3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8115386-76d1-40b4-a5e7-a597e133c3bc",
   "metadata": {},
   "source": [
    "# Isolate sequencing\n",
    "\n",
    "Same steps as for metagenomics. However, minimum allele frequency is set to 0.95.\n",
    "\n",
    "## File- and directory structure\n",
    "\n",
    "The data is on `/work/FAC/FBM/DMF/smitri/default/leakage/data/isolate_sequencing`. Code in `/work/FAC/FBM/DMF/smitri/default/leakage/code/isolate_sequncing`. I expotred the variants using the `vcf_exporter.py` in `../../code/isolate_sequencing/results/vcf_exporter.py`.  \n",
    "The filtered variants with frequency >= 0.95 are in `../../code/isolate_sequencing/results/filtered_variants`. All variants are in `../../code/isolate_sequencing/results/all_variants`.\n",
    "aceE_samples = ['AceE_M2_D44','AceE_M3_D44','AceE_M4_D44']\n",
    "sucB_samples = ['SucB_M5','SucB_M6','SucB_M7']\n",
    "all_samples = aceE_samples + sucB_samples\n",
    "dfs = [pd.read_csv('../../code/meta_sequencing/results/all_filtered_variants/' + sample + '.filtered.csv') for sample in samples]\n",
    "\n",
    "## SNPs\n",
    "\n",
    "Bad news, *ynfM* is in the ancestor as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "577161b0-7437-43da-a599-3da6aff805bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>gene</th>\n",
       "      <th>product</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1455</td>\n",
       "      <td>thrA</td>\n",
       "      <td>Bifunctional aspartokinase/homoserine dehydrog...</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2297</td>\n",
       "      <td>thrA</td>\n",
       "      <td>Bifunctional aspartokinase/homoserine dehydrog...</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9100</td>\n",
       "      <td>talB</td>\n",
       "      <td>transaldolase B</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9125</td>\n",
       "      <td>talB</td>\n",
       "      <td>transaldolase B</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>4532543</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>4538883</td>\n",
       "      <td>fimH</td>\n",
       "      <td>minor component of type 1 fimbriae</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>4559788</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>4579955</td>\n",
       "      <td>yjiY</td>\n",
       "      <td>putative transporter</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>4596332</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>270 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         pos  gene                                            product  \\\n",
       "0       1455  thrA  Bifunctional aspartokinase/homoserine dehydrog...   \n",
       "1       2297  thrA  Bifunctional aspartokinase/homoserine dehydrog...   \n",
       "2       9100  talB                                    transaldolase B   \n",
       "3       9125  talB                                    transaldolase B   \n",
       "4      22360   NaN                                                NaN   \n",
       "..       ...   ...                                                ...   \n",
       "265  4532543   NaN                                                NaN   \n",
       "266  4538883  fimH                 minor component of type 1 fimbriae   \n",
       "267  4559788   NaN                                                NaN   \n",
       "268  4579955  yjiY                               putative transporter   \n",
       "269  4596332   NaN                                                NaN   \n",
       "\n",
       "          sample  \n",
       "0    AceE_M4_iso  \n",
       "1    AceE_M4_iso  \n",
       "2    AceE_M4_iso  \n",
       "3    AceE_M4_iso  \n",
       "4    AceE_M4_iso  \n",
       "..           ...  \n",
       "265  AceE_M4_iso  \n",
       "266  AceE_M4_iso  \n",
       "267  AceE_M4_iso  \n",
       "268  AceE_M4_iso  \n",
       "269  AceE_M4_iso  \n",
       "\n",
       "[270 rows x 4 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aceE_samples = ['AceE_anc','AceE_M2_iso','AceE_M3_iso','AceE_M4_iso']\n",
    "sucB_samples = ['SucB_anc','SucB_M5_iso','SucB_M6_iso','SucB_M7_iso']\n",
    "all_samples = aceE_samples + sucB_samples\n",
    "dfs = [pd.read_csv('../../code/isolate_sequncing/results/filtered_variants/' + sample + '.filtered.csv')[['pos','gene','product','sample']] for sample in aceE_samples]\n",
    "dfs[3]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
