{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9d706b18",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6432064",
   "metadata": {},
   "source": [
    "# Read in the variant analysis result for the isolate sequencing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3a59a873",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Define the folder path\n",
    "data_path = Path(\"../../code/isolate_sequencing/results/filtered_variants\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "de7a5c19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List all CSV files in the folder\n",
    "csv_files = [file for file in data_path.glob('*.csv')]\n",
    "# Read each CSV file into a dictionary of DataFrames\n",
    "dataframes = {file.stem: pd.read_csv(file) for file in csv_files}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9b08cc21",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# name_to_file = {x.split('.')[0].replace('_iso','').replace('_anc', ''): x for x in dataframes.keys()}\n",
    "name_to_file = {\n",
    "    'sucB-M6': 'SucB_M6_iso.filtered',\n",
    "    'aceE': 'AceE_anc.filtered',\n",
    "    'sucB': 'SucB_anc.filtered',\n",
    "    'aceE-M4': 'AceE_M4_iso.filtered',\n",
    "    'sucB-M7': 'SucB_M7_iso.filtered',\n",
    "    'aceE-M3': 'AceE_M3_iso.filtered',\n",
    "    'all_samples': 'all_samples.filtered',\n",
    "    'aceE-M2': 'AceE_M2_iso.filtered',\n",
    "    'sucB-M5': 'SucB_M5_iso.filtered'}\n",
    "aceE_evolved = ['aceE-M2', 'aceE-M3', 'aceE-M4']\n",
    "sucB_evolved = ['sucB-M5', 'sucB-M6', 'sucB-M7']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57223b5f",
   "metadata": {},
   "source": [
    "## Remove the SNPs that are also in the ancestors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a095723d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing aceE-M2 row 1 with pos 1622605\n",
      "Removing aceE-M2 row 2 with pos 1664690\n",
      "Removing aceE-M2 row 4 with pos 2719426\n",
      "Removing aceE-M3 row 1 with pos 1622605\n",
      "Removing aceE-M3 row 2 with pos 1664690\n",
      "Removing aceE-M3 row 3 with pos 2719426\n",
      "Removing aceE-M4 row 96 with pos 1622605\n",
      "Removing aceE-M4 row 100 with pos 1664690\n",
      "Removing aceE-M4 row 162 with pos 2719426\n",
      "Removing sucB-M5 row 0 with pos 447145\n",
      "Removing sucB-M5 row 2 with pos 2719426\n",
      "Removing sucB-M6 row 0 with pos 447145\n",
      "Removing sucB-M6 row 1 with pos 2719426\n",
      "Removing sucB-M7 row 0 with pos 447145\n",
      "Removing sucB-M7 row 3 with pos 2719426\n"
     ]
    }
   ],
   "source": [
    "# aceE\n",
    "aceE_df = dataframes[name_to_file['aceE']]\n",
    "aceE_locations = aceE_df['pos'].unique()\n",
    "for isolate in aceE_evolved:\n",
    "    isolate_df = dataframes[name_to_file[isolate]]\n",
    "    rows_to_remove = []\n",
    "    for i, row in isolate_df.iterrows():\n",
    "        if row['pos'] in aceE_locations:\n",
    "            print(f\"Removing {isolate} row {i} with pos {row['pos']}\")\n",
    "            rows_to_remove.append(i)\n",
    "    isolate_df.drop(rows_to_remove, inplace=True)\n",
    "\n",
    "sucB_df = dataframes[name_to_file['sucB']]\n",
    "sucB_locations = sucB_df['pos'].unique()\n",
    "for isolate in sucB_evolved:\n",
    "    isolate_df = dataframes[name_to_file[isolate]]\n",
    "    rows_to_remove = []\n",
    "    for i, row in isolate_df.iterrows():\n",
    "        if row['pos'] in sucB_locations:\n",
    "            print(f\"Removing {isolate} row {i} with pos {row['pos']}\")\n",
    "            rows_to_remove.append(i)\n",
    "    isolate_df.drop(rows_to_remove, inplace=True)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "46a83df8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(267, 15)\n"
     ]
    }
   ],
   "source": [
    "ace_M4_df = dataframes[name_to_file['aceE-M4']]\n",
    "print(ace_M4_df.shape)\n",
    "m4_genes = ace_M4_df.gene.unique()\n",
    "pd.Series(m4_genes).to_csv(data_path / 'm4_genes.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e59847b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>qual</th>\n",
       "      <th>depth</th>\n",
       "      <th>freq</th>\n",
       "      <th>alt</th>\n",
       "      <th>alt_count</th>\n",
       "      <th>ref</th>\n",
       "      <th>type</th>\n",
       "      <th>len</th>\n",
       "      <th>eff</th>\n",
       "      <th>gene</th>\n",
       "      <th>product</th>\n",
       "      <th>linegroup</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>3081560</td>\n",
       "      <td>6450.74</td>\n",
       "      <td>186</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>186</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CP009273.3081560.SucB_M6_iso</td>\n",
       "      <td>SucB_M6_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>3782776</td>\n",
       "      <td>6329.78</td>\n",
       "      <td>179</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>179</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>waaH</td>\n",
       "      <td>LPS(HepIII)-glucuronic acid glycosyltransferase</td>\n",
       "      <td>CP009273.3782776.SucB_M6_iso</td>\n",
       "      <td>SucB_M6_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>4209467</td>\n",
       "      <td>7773.93</td>\n",
       "      <td>220</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>220</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>aceK</td>\n",
       "      <td>isocitrate dehydrogenase kinase/phosphatase</td>\n",
       "      <td>CP009273.4209467.SucB_M6_iso</td>\n",
       "      <td>SucB_M6_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>4538904</td>\n",
       "      <td>7974.41</td>\n",
       "      <td>235</td>\n",
       "      <td>0.995745</td>\n",
       "      <td>Substitution(type_='SNV', value='T')</td>\n",
       "      <td>234</td>\n",
       "      <td>G</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>fimH</td>\n",
       "      <td>minor component of type 1 fimbriae</td>\n",
       "      <td>CP009273.4538904.SucB_M6_iso</td>\n",
       "      <td>SucB_M6_iso</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      pos     qual  depth      freq  \\\n",
       "2  CP009273  3081560  6450.74    186  1.000000   \n",
       "3  CP009273  3782776  6329.78    179  1.000000   \n",
       "4  CP009273  4209467  7773.93    220  1.000000   \n",
       "5  CP009273  4538904  7974.41    235  0.995745   \n",
       "\n",
       "                                    alt  alt_count ref type  len  \\\n",
       "2  Substitution(type_='SNV', value='A')        186   C  snp    1   \n",
       "3  Substitution(type_='SNV', value='A')        179   C  snp    1   \n",
       "4  Substitution(type_='SNV', value='A')        220   C  snp    1   \n",
       "5  Substitution(type_='SNV', value='T')        234   G  snp    1   \n",
       "\n",
       "                  eff  gene                                          product  \\\n",
       "2                 NaN   NaN                                              NaN   \n",
       "3  synonymous_variant  waaH  LPS(HepIII)-glucuronic acid glycosyltransferase   \n",
       "4    missense_variant  aceK      isocitrate dehydrogenase kinase/phosphatase   \n",
       "5    missense_variant  fimH               minor component of type 1 fimbriae   \n",
       "\n",
       "                      linegroup       sample  \n",
       "2  CP009273.3081560.SucB_M6_iso  SucB_M6_iso  \n",
       "3  CP009273.3782776.SucB_M6_iso  SucB_M6_iso  \n",
       "4  CP009273.4209467.SucB_M6_iso  SucB_M6_iso  \n",
       "5  CP009273.4538904.SucB_M6_iso  SucB_M6_iso  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes[name_to_file['sucB-M6']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fb18e1c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>qual</th>\n",
       "      <th>depth</th>\n",
       "      <th>freq</th>\n",
       "      <th>alt</th>\n",
       "      <th>alt_count</th>\n",
       "      <th>ref</th>\n",
       "      <th>type</th>\n",
       "      <th>len</th>\n",
       "      <th>eff</th>\n",
       "      <th>gene</th>\n",
       "      <th>product</th>\n",
       "      <th>linegroup</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>752345</td>\n",
       "      <td>6971.83</td>\n",
       "      <td>198</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Substitution(type_='SNV', value='T')</td>\n",
       "      <td>198</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>succinate dehydrogenase, flavoprotein subunit</td>\n",
       "      <td>CP009273.752345.SucB_M7_iso</td>\n",
       "      <td>SucB_M7_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>2234612</td>\n",
       "      <td>3502.85</td>\n",
       "      <td>112</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Substitution(type_='INDEL', value='AG')</td>\n",
       "      <td>112</td>\n",
       "      <td>ATACGTTGATG</td>\n",
       "      <td>del</td>\n",
       "      <td>9</td>\n",
       "      <td>disruptive_inframe_deletion</td>\n",
       "      <td>galS</td>\n",
       "      <td>galactose- and fucose-inducible galactose regu...</td>\n",
       "      <td>CP009273.2234612.SucB_M7_iso</td>\n",
       "      <td>SucB_M7_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>3782776</td>\n",
       "      <td>8405.95</td>\n",
       "      <td>238</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>238</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>waaH</td>\n",
       "      <td>LPS(HepIII)-glucuronic acid glycosyltransferase</td>\n",
       "      <td>CP009273.3782776.SucB_M7_iso</td>\n",
       "      <td>SucB_M7_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>4209467</td>\n",
       "      <td>7904.26</td>\n",
       "      <td>224</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>224</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>aceK</td>\n",
       "      <td>isocitrate dehydrogenase kinase/phosphatase</td>\n",
       "      <td>CP009273.4209467.SucB_M7_iso</td>\n",
       "      <td>SucB_M7_iso</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      pos     qual  depth  freq  \\\n",
       "1  CP009273   752345  6971.83    198   1.0   \n",
       "2  CP009273  2234612  3502.85    112   1.0   \n",
       "4  CP009273  3782776  8405.95    238   1.0   \n",
       "5  CP009273  4209467  7904.26    224   1.0   \n",
       "\n",
       "                                       alt  alt_count          ref type  len  \\\n",
       "1     Substitution(type_='SNV', value='T')        198            C  snp    1   \n",
       "2  Substitution(type_='INDEL', value='AG')        112  ATACGTTGATG  del    9   \n",
       "4     Substitution(type_='SNV', value='A')        238            C  snp    1   \n",
       "5     Substitution(type_='SNV', value='A')        224            C  snp    1   \n",
       "\n",
       "                           eff  gene  \\\n",
       "1             missense_variant  sdhA   \n",
       "2  disruptive_inframe_deletion  galS   \n",
       "4           synonymous_variant  waaH   \n",
       "5             missense_variant  aceK   \n",
       "\n",
       "                                             product  \\\n",
       "1      succinate dehydrogenase, flavoprotein subunit   \n",
       "2  galactose- and fucose-inducible galactose regu...   \n",
       "4    LPS(HepIII)-glucuronic acid glycosyltransferase   \n",
       "5        isocitrate dehydrogenase kinase/phosphatase   \n",
       "\n",
       "                      linegroup       sample  \n",
       "1   CP009273.752345.SucB_M7_iso  SucB_M7_iso  \n",
       "2  CP009273.2234612.SucB_M7_iso  SucB_M7_iso  \n",
       "4  CP009273.3782776.SucB_M7_iso  SucB_M7_iso  \n",
       "5  CP009273.4209467.SucB_M7_iso  SucB_M7_iso  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes[name_to_file['sucB-M7']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a278b8d0",
   "metadata": {},
   "source": [
    "# Export all isolate data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "00512812",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>qual</th>\n",
       "      <th>depth</th>\n",
       "      <th>freq</th>\n",
       "      <th>alt</th>\n",
       "      <th>alt_count</th>\n",
       "      <th>ref</th>\n",
       "      <th>type</th>\n",
       "      <th>len</th>\n",
       "      <th>eff</th>\n",
       "      <th>gene</th>\n",
       "      <th>product</th>\n",
       "      <th>linegroup</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>904878</td>\n",
       "      <td>7494.39</td>\n",
       "      <td>214</td>\n",
       "      <td>0.995327</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>213</td>\n",
       "      <td>G</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>poxB</td>\n",
       "      <td>pyruvate dehydrogenase (pyruvate oxidase), thi...</td>\n",
       "      <td>CP009273.904878.AceE_M2_iso</td>\n",
       "      <td>AceE_M2_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>2235029</td>\n",
       "      <td>5943.62</td>\n",
       "      <td>168</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Substitution(type_='SNV', value='C')</td>\n",
       "      <td>168</td>\n",
       "      <td>A</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>galS</td>\n",
       "      <td>galactose- and fucose-inducible galactose regu...</td>\n",
       "      <td>CP009273.2235029.AceE_M2_iso</td>\n",
       "      <td>AceE_M2_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>3528650</td>\n",
       "      <td>6552.52</td>\n",
       "      <td>186</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>186</td>\n",
       "      <td>G</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>envZ</td>\n",
       "      <td>sensory histidine kinase in two-component regu...</td>\n",
       "      <td>CP009273.3528650.AceE_M2_iso</td>\n",
       "      <td>AceE_M2_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>3574489</td>\n",
       "      <td>6674.87</td>\n",
       "      <td>189</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Substitution(type_='SNV', value='T')</td>\n",
       "      <td>189</td>\n",
       "      <td>G</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CP009273.3574489.AceE_M2_iso</td>\n",
       "      <td>AceE_M2_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>696432</td>\n",
       "      <td>7277.51</td>\n",
       "      <td>209</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>209</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>nagC</td>\n",
       "      <td>N-acetylglucosamine-inducible nag divergent op...</td>\n",
       "      <td>CP009273.696432.AceE_M3_iso</td>\n",
       "      <td>AceE_M3_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>4532543</td>\n",
       "      <td>7297.24</td>\n",
       "      <td>206</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Substitution(type_='SNV', value='T')</td>\n",
       "      <td>206</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CP009273.4532543.AceE_M4_iso</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>4538883</td>\n",
       "      <td>8557.67</td>\n",
       "      <td>246</td>\n",
       "      <td>0.995935</td>\n",
       "      <td>Substitution(type_='SNV', value='T')</td>\n",
       "      <td>245</td>\n",
       "      <td>G</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>fimH</td>\n",
       "      <td>minor component of type 1 fimbriae</td>\n",
       "      <td>CP009273.4538883.AceE_M4_iso</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>4559788</td>\n",
       "      <td>8907.80</td>\n",
       "      <td>258</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Substitution(type_='SNV', value='T')</td>\n",
       "      <td>258</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CP009273.4559788.AceE_M4_iso</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>4579955</td>\n",
       "      <td>9584.67</td>\n",
       "      <td>275</td>\n",
       "      <td>0.992727</td>\n",
       "      <td>Substitution(type_='SNV', value='C')</td>\n",
       "      <td>273</td>\n",
       "      <td>A</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>yjiY</td>\n",
       "      <td>putative transporter</td>\n",
       "      <td>CP009273.4579955.AceE_M4_iso</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>4596332</td>\n",
       "      <td>7189.11</td>\n",
       "      <td>203</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>203</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CP009273.4596332.AceE_M4_iso</td>\n",
       "      <td>AceE_M4_iso</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>273 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      pos     qual  depth      freq  \\\n",
       "0    CP009273   904878  7494.39    214  0.995327   \n",
       "3    CP009273  2235029  5943.62    168  1.000000   \n",
       "5    CP009273  3528650  6552.52    186  1.000000   \n",
       "6    CP009273  3574489  6674.87    189  1.000000   \n",
       "0    CP009273   696432  7277.51    209  1.000000   \n",
       "..        ...      ...      ...    ...       ...   \n",
       "265  CP009273  4532543  7297.24    206  1.000000   \n",
       "266  CP009273  4538883  8557.67    246  0.995935   \n",
       "267  CP009273  4559788  8907.80    258  1.000000   \n",
       "268  CP009273  4579955  9584.67    275  0.992727   \n",
       "269  CP009273  4596332  7189.11    203  1.000000   \n",
       "\n",
       "                                      alt  alt_count ref type  len  \\\n",
       "0    Substitution(type_='SNV', value='A')        213   G  snp    1   \n",
       "3    Substitution(type_='SNV', value='C')        168   A  snp    1   \n",
       "5    Substitution(type_='SNV', value='A')        186   G  snp    1   \n",
       "6    Substitution(type_='SNV', value='T')        189   G  snp    1   \n",
       "0    Substitution(type_='SNV', value='A')        209   C  snp    1   \n",
       "..                                    ...        ...  ..  ...  ...   \n",
       "265  Substitution(type_='SNV', value='T')        206   C  snp    1   \n",
       "266  Substitution(type_='SNV', value='T')        245   G  snp    1   \n",
       "267  Substitution(type_='SNV', value='T')        258   C  snp    1   \n",
       "268  Substitution(type_='SNV', value='C')        273   A  snp    1   \n",
       "269  Substitution(type_='SNV', value='A')        203   C  snp    1   \n",
       "\n",
       "                  eff  gene  \\\n",
       "0    missense_variant  poxB   \n",
       "3    missense_variant  galS   \n",
       "5    missense_variant  envZ   \n",
       "6                 NaN   NaN   \n",
       "0    missense_variant  nagC   \n",
       "..                ...   ...   \n",
       "265               NaN   NaN   \n",
       "266  missense_variant  fimH   \n",
       "267               NaN   NaN   \n",
       "268  missense_variant  yjiY   \n",
       "269               NaN   NaN   \n",
       "\n",
       "                                               product  \\\n",
       "0    pyruvate dehydrogenase (pyruvate oxidase), thi...   \n",
       "3    galactose- and fucose-inducible galactose regu...   \n",
       "5    sensory histidine kinase in two-component regu...   \n",
       "6                                                  NaN   \n",
       "0    N-acetylglucosamine-inducible nag divergent op...   \n",
       "..                                                 ...   \n",
       "265                                                NaN   \n",
       "266                 minor component of type 1 fimbriae   \n",
       "267                                                NaN   \n",
       "268                               putative transporter   \n",
       "269                                                NaN   \n",
       "\n",
       "                        linegroup       sample  \n",
       "0     CP009273.904878.AceE_M2_iso  AceE_M2_iso  \n",
       "3    CP009273.2235029.AceE_M2_iso  AceE_M2_iso  \n",
       "5    CP009273.3528650.AceE_M2_iso  AceE_M2_iso  \n",
       "6    CP009273.3574489.AceE_M2_iso  AceE_M2_iso  \n",
       "0     CP009273.696432.AceE_M3_iso  AceE_M3_iso  \n",
       "..                            ...          ...  \n",
       "265  CP009273.4532543.AceE_M4_iso  AceE_M4_iso  \n",
       "266  CP009273.4538883.AceE_M4_iso  AceE_M4_iso  \n",
       "267  CP009273.4559788.AceE_M4_iso  AceE_M4_iso  \n",
       "268  CP009273.4579955.AceE_M4_iso  AceE_M4_iso  \n",
       "269  CP009273.4596332.AceE_M4_iso  AceE_M4_iso  \n",
       "\n",
       "[273 rows x 15 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aceE_isolates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "2cf0d81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "aceE_isolates = pd.concat([dataframes[name_to_file[isolate]] for isolate in aceE_evolved])\n",
    "sucB_isolates = pd.concat([dataframes[name_to_file[isolate]] for isolate in sucB_evolved])\n",
    "\n",
    "aceE_isolates[['sample', 'pos', 'chrom', 'gene','freq', 'type', 'len', 'eff', 'product']].to_csv(data_path.parent / 'aceE_variants.csv')\n",
    "sucB_isolates[['sample', 'pos', 'chrom', 'gene','freq', 'type', 'len', 'eff', 'product']].to_csv(data_path.parent / 'sucB_variants.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c730efa",
   "metadata": {},
   "source": [
    "# Read in the variants from the meta sequencing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6ffd46b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_folder = Path('../../code/meta_sequencing/results/fixed_filtered_variants')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b701c2ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List all CSV files in the folder\n",
    "meta_csv_files = [file for file in meta_folder.glob('*.csv')]\n",
    "# Read each CSV file into a dictionary of DataFrames\n",
    "meta_dataframes = {file.stem: pd.read_csv(file) for file in meta_csv_files}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "de5c4ca1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['SucB_M6.filtered', 'AceE_M4_D44.filtered', 'SucB_M7.filtered', 'AceE_M3_D44.filtered', 'SucB_M5.filtered', 'all_samples.filtered', 'AceE_M2_D44.filtered'])"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_dataframes.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "46cab721",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_meta_df = meta_dataframes['all_samples.filtered']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4b2d7877",
   "metadata": {},
   "outputs": [],
   "source": [
    "aceE_all = all_meta_df.loc[all_meta_df['sample'].str.contains('AceE'), :]\n",
    "sucB_all = all_meta_df.loc[all_meta_df['sample'].str.contains('SucB'), :]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e62f0fb",
   "metadata": {},
   "source": [
    "# Remove SNPs also in the ancestor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "9b06fe54",
   "metadata": {},
   "outputs": [],
   "source": [
    "#aceE\n",
    "aceE_df = dataframes[name_to_file['aceE']]\n",
    "aceE_locations = aceE_df['pos'].unique()\n",
    "for i, row in aceE_all.iterrows():\n",
    "    if row['pos'] in aceE_locations:\n",
    "        print(f\"Removing {row['sample']} row {i} with pos {row['pos']}, gene {row['gene']}\")\n",
    "        aceE_all.drop(i, inplace=True)\n",
    "\n",
    "sucB_df = dataframes[name_to_file['sucB']]\n",
    "sucB_locations = sucB_df['pos'].unique()\n",
    "for i, row in sucB_all.iterrows():\n",
    "    if row['pos'] in sucB_locations:\n",
    "        print(f\"Removing {row['sample']} row {i} with pos {row['pos']}, gene {row['gene']}\")\n",
    "        sucB_all.drop(i, inplace=True)\n",
    "\n",
    "        \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6901faeb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "eea654b2",
   "metadata": {},
   "source": [
    "# Export spreadsheet with the most relevant data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "6336d194",
   "metadata": {},
   "outputs": [],
   "source": [
    "sucB_all_reduced = sucB_all[['sample', 'pos', 'chrom', 'gene', 'freq', 'ref', 'alt', 'len', 'type','eff', 'product']]\n",
    "sucB_all_reduced.to_csv(meta_folder.parent / 'sucB_all_fixed_filtered.csv', index=False)\n",
    "aceE_all_reduced = aceE_all[['sample', 'pos', 'chrom', 'gene', 'freq', 'ref', 'alt', 'len', 'type','eff', 'product']]\n",
    "aceE_all_reduced.to_csv(meta_folder.parent / 'aceE_all_fixed_filtered.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5965dc8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>pos</th>\n",
       "      <th>gene</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>eff</th>\n",
       "      <th>product</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sample, pos, gene, ref, alt, eff, product]\n",
       "Index: []"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sucB_all_reduced.loc[sucB_all_reduced['sample'].str.contains('M5')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a3711e43",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[nan 'ydbD' 'galS' 'waaH' 'aceK' 'ldhA' 'ynfM' 'ilvG']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "unique_duplicated_genes = all_meta_df['gene'][all_meta_df['gene'].duplicated()].unique()\n",
    "\n",
    "# Display the unique duplicated genes\n",
    "print(unique_duplicated_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8cebd3b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>qual</th>\n",
       "      <th>depth</th>\n",
       "      <th>freq</th>\n",
       "      <th>alt</th>\n",
       "      <th>alt_count</th>\n",
       "      <th>ref</th>\n",
       "      <th>type</th>\n",
       "      <th>len</th>\n",
       "      <th>eff</th>\n",
       "      <th>gene</th>\n",
       "      <th>product</th>\n",
       "      <th>linegroup</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>3782776</td>\n",
       "      <td>69202.2</td>\n",
       "      <td>1980</td>\n",
       "      <td>0.991414</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>1963</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>waaH</td>\n",
       "      <td>LPS(HepIII)-glucuronic acid glycosyltransferase</td>\n",
       "      <td>CP009273.3782776.SucB_M6</td>\n",
       "      <td>SucB_M6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>3782776</td>\n",
       "      <td>92280.5</td>\n",
       "      <td>2659</td>\n",
       "      <td>0.987965</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>2627</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>waaH</td>\n",
       "      <td>LPS(HepIII)-glucuronic acid glycosyltransferase</td>\n",
       "      <td>CP009273.3782776.SucB_M7</td>\n",
       "      <td>SucB_M7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      pos     qual  depth      freq  \\\n",
       "103  CP009273  3782776  69202.2   1980  0.991414   \n",
       "134  CP009273  3782776  92280.5   2659  0.987965   \n",
       "\n",
       "                                      alt  alt_count ref type  len  \\\n",
       "103  Substitution(type_='SNV', value='A')       1963   C  snp    1   \n",
       "134  Substitution(type_='SNV', value='A')       2627   C  snp    1   \n",
       "\n",
       "                    eff  gene  \\\n",
       "103  synonymous_variant  waaH   \n",
       "134  synonymous_variant  waaH   \n",
       "\n",
       "                                             product  \\\n",
       "103  LPS(HepIII)-glucuronic acid glycosyltransferase   \n",
       "134  LPS(HepIII)-glucuronic acid glycosyltransferase   \n",
       "\n",
       "                    linegroup   sample  \n",
       "103  CP009273.3782776.SucB_M6  SucB_M6  \n",
       "134  CP009273.3782776.SucB_M7  SucB_M7  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sucB_all.loc[sucB_all.gene=='waaH'].head(40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "aa2a7f30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([nan, 'ydbD', 'galS', 'waaH', 'aceK'], dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sucB_all['gene'][sucB_all['gene'].duplicated()].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "795c649b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.Series(sucB_all.gene.unique()).to_csv(meta_folder / 'sucB_genes.csv', index=False)\n",
    "pd.Series(aceE_all.gene.unique()).to_csv(meta_folder / 'aceE_genes.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "a92f901e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>qual</th>\n",
       "      <th>depth</th>\n",
       "      <th>freq</th>\n",
       "      <th>alt</th>\n",
       "      <th>alt_count</th>\n",
       "      <th>ref</th>\n",
       "      <th>type</th>\n",
       "      <th>len</th>\n",
       "      <th>eff</th>\n",
       "      <th>gene</th>\n",
       "      <th>product</th>\n",
       "      <th>linegroup</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>1470490</td>\n",
       "      <td>57737.3</td>\n",
       "      <td>1705</td>\n",
       "      <td>0.980059</td>\n",
       "      <td>Substitution(type_='SNV', value='A')</td>\n",
       "      <td>1671</td>\n",
       "      <td>C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ydbD</td>\n",
       "      <td>PF10971 family putative periplasmic methylglyo...</td>\n",
       "      <td>CP009273.1470490.SucB_M6</td>\n",
       "      <td>SucB_M6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>CP009273</td>\n",
       "      <td>1471105</td>\n",
       "      <td>62035.6</td>\n",
       "      <td>1818</td>\n",
       "      <td>0.981848</td>\n",
       "      <td>Substitution(type_='SNV', value='C')</td>\n",
       "      <td>1785</td>\n",
       "      <td>A</td>\n",
       "      <td>snp</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>ydbD</td>\n",
       "      <td>PF10971 family putative periplasmic methylglyo...</td>\n",
       "      <td>CP009273.1471105.SucB_M6</td>\n",
       "      <td>SucB_M6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom      pos     qual  depth      freq  \\\n",
       "47  CP009273  1470490  57737.3   1705  0.980059   \n",
       "48  CP009273  1471105  62035.6   1818  0.981848   \n",
       "\n",
       "                                     alt  alt_count ref type  len  \\\n",
       "47  Substitution(type_='SNV', value='A')       1671   C  snp    1   \n",
       "48  Substitution(type_='SNV', value='C')       1785   A  snp    1   \n",
       "\n",
       "                 eff  gene                                            product  \\\n",
       "47  missense_variant  ydbD  PF10971 family putative periplasmic methylglyo...   \n",
       "48  missense_variant  ydbD  PF10971 family putative periplasmic methylglyo...   \n",
       "\n",
       "                   linegroup   sample  \n",
       "47  CP009273.1470490.SucB_M6  SucB_M6  \n",
       "48  CP009273.1471105.SucB_M6  SucB_M6  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_meta_df.loc[all_meta_df['gene'] == 'ydbD']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
